lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -42,6 +106,135 @@ def urlencode_json(json: JSON) -> str:
+
+    def __init__(self, term: str):
+        self.chain = None
+        self.key = None
+        self.value = None
+        self.term = term
+        self.buffer = term
+        self.logger = logging.getLogger("term")
+
+    def _remove_prefix(
+        self, prefixes: tuple[str], default: Optional[str] = None
+    ) -> Optional[str]:
+        """Remove a string prefix.
+
+        Looking at the current buffer, remove a known prefix before processing
+        the next character. A list of prefixes may be specified: they're sorted
+        longest first so that we don't accidentally match a substring.
+
+        Args:
+            prefixes: a collection of prefix strings
+            default: the default prefix if none appears in the string.
+
+        Returns:
+            The prefix, or the default.
+        """
+        prefix = default
+        for p in sorted(prefixes, key=lambda k: len(k), reverse=True):
+            if self.buffer.startswith(p):
+                prefix = self.buffer[: len(p)]
+                self.buffer = self.buffer[len(p) :]
+                break
+        return prefix
+
+    def _next_token(self, optional: bool = False) -> str:
+        """Extract a string from the head of the buffer.
+
+        To filter using metadata keys (e.g., unusual benchmark generated
+        metadata.log keys) or values (e.g. a date/time string) which may
+        contain ":" symbols, the key or value may be enclosed in quotes.
+        Either single or double quotes are allowed.
+
+        We're looking for the ":" symbol delimiting the next term of the filter
+        expression, or (if that's marked "optional") the end of the buffer.
+
+        Args:
+            optional: The terminating ":" separator is optional.
+
+        Returns:
+            The token, unquoted
+        """
+        buffer = self.buffer
+        if buffer.startswith(("'", '"')):
+            quote = buffer[:1]
+            end = buffer.find(quote, 1)
+            if end < 0:
+                raise APIAbort(
+                    HTTPStatus.BAD_REQUEST, f"Bad quote termination in {self.term!r}"
+                )
+            next = buffer[1:end]
+            end += 1
+        else:
+            end = buffer.find(":")
+            if end < 0:
+                if not optional:
+                    raise APIAbort(
+                        HTTPStatus.BAD_REQUEST, f"Missing terminator for {buffer!r}"
+                    )
+                next = buffer
+                end = len(buffer)
+            else:
+                next = buffer[:end]
+        buffer = buffer[end:]
+        if not optional and (len(buffer) == 0 or not buffer.startswith(":")):
+            self.logger.info(
+                "[optional %s, len %d, next %s]", optional, len(buffer), buffer[0]
+            )
+            raise APIAbort(HTTPStatus.BAD_REQUEST, f"Missing terminator for {buffer!r}")
+        buffer = buffer[1:]
+        self.buffer = buffer
+        return next
```

Comment:
At line 174-177 are we only logging the error and not terminating the whole process?

Reply:
We're not logging anything here: we generally don't log user errors, because it could fill up our logs without any real benefit. We're raising an `APIAbort` that will fail the API call and report a `response.json["message"]` explanation back to the client.